### PR TITLE
[core] propagate oom exception when worker is killed due to oom

### PR DESCRIPTION
**Pull request by clarng**
Signed-off-by: Clarence Ng <clarence.wyng@gmail.com>

## Why are these changes needed?

When task fails we currently report a worker crash error and there is no way to customize the message. This PR introduces the ability to report a custom error with additional information on why the task failed. 

This PR is also necessary in order for us to perform custom retry when the task failed due to worker OOM.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

Updated error message in the logs and the exception

OOM error:

```
E                       ray.exceptions.OutOfMemoryError: Task was killed due to the node running low on memory.
E                       
E                       Memory on the node (IP: 172.17.0.2, ID: 974c4ed2f1c84f6ec55e7ff409107e2184310c0e8e6bdb048b9b7f5e) where the task was running was 23.56GB / 33.28GB (0.707765), which exceeds the memory usage threshold of 0.7. Ray killed this worker (ID: caedf7cd60551e06e2a4b12d921f02eaa1e949201730ac23eb80db8f) because it was the most recently scheduled task; to see more information about memory usage on this node, use `ray logs raylet.out -ip 172.17.0.2`. To see the logs of the worker, use `ray logs worker-caedf7cd60551e06e2a4b12d921f02eaa1e949201730ac23eb80db8f*out -ip 172.17.0.2`.
E                       
E                       Consider provisioning more memory on this node or reducing task parallelism by requesting more CPUs per task. To adjust the eviction threshold, set the environment variable `RAY_memory_usage_threshold_fraction` when starting Ray. To disable worker eviction, set the environment variable `RAY_memory_monitor_interval_ms` to zero.
```

Node died error:
```
E                       ray.exceptions.NodeDiedError: Task failed due to the node dying.
E                       
E                       The node (IP: 172.17.0.2, node ID: a98894c0b41cf82f5ddf16f3869dd982d068677675f5b0bd8a89e8b6) where this task was running crashed unexpectedly. This can happen if: (1) the instance where the node was running failed, (2) raylet crashes unexpectedly (OOM, preempted node, etc). To see more information about the crash, use `ray logs raylet.out -ip 172.17.0.2`
```

### CODE REVIEW

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (stephanie-wang) [inline src/ray/core_worker/transport/direct_task_transport.cc:593]: I was a bit confused about this part. Why do we need to check the thread ID?

File: src/ray/raylet/node_manager.cc
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.cc:3020]: When/how should we GC these?

**Review by stephanie-wang — COMMENTED** [review]
Looks pretty good so far!

File: src/ray/raylet/node_manager.cc
Comment (clarng) [reply src/ray/raylet/node_manager.cc:3020]: Chatted offline seems we could consider writing those records to GCS when the entry expires on the raylet, @rkooo567 thoughts?

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (clarng) [reply src/ray/core_worker/transport/direct_task_transport.cc:593]: mostly being defensive, since currently this code runs inline, this verifies it is still using the same thread (and in general it seems that's the current threading model) 

Can remove if we don't want to have these checks, let me know your preference?

Comment (rkooo567) [inline src/ray/core_worker/transport/direct_task_transport.cc:593]: IIUC, it is not guaranteed? (the PushNormalTask is executed in the main thread and the callback is executed inside io thread?)

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (rkooo567) [inline src/ray/core_worker/transport/direct_task_transport.cc:594]: I was confused by this at first. Can you add a comment like this

```
RAY_CHECK(!status.ok()) << "Checking failed task status RPC was called although the test didn't fail";
```

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (rkooo567) [inline src/ray/core_worker/transport/direct_task_transport.cc:602]: is it TODO in this PR? I think it is pretty easy to add it here. We can also add more metadata like node id

File: src/ray/raylet/node_manager.cc
Comment (rkooo567) [inline src/ray/raylet/node_manager.cc:825]: Does it also cover cases like worker failure errors?

File: src/ray/raylet/node_manager.cc
Comment (rkooo567) [reply src/ray/raylet/node_manager.cc:3020]: maybe we should record it on the owner side since we query the owner for the state API?

File: src/ray/raylet/node_manager.cc
Comment (rkooo567) [inline src/ray/raylet/node_manager.cc:2996]: Looks like we only take care of OOM For this now? are you planning to handle other error types? (like worker failure)

**Review by rkooo567 — COMMENTED** [review]
Looks like we only handle OOM case in this PR. Are you also planning to follow up with other types of failures (e.g., worker failure)?

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (clarng) [reply src/ray/core_worker/transport/direct_task_transport.cc:602]: do you mean we log it as node failure if the rpc failed?

File: src/ray/raylet/node_manager.cc
Comment (clarng) [reply src/ray/raylet/node_manager.cc:2996]: Can we do that in a follow up PR or has to be here? Just worried it will take time away to ensure it is done properly

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (clarng) [reply src/ray/core_worker/transport/direct_task_transport.cc:593]: confirmed offline this all opens on io thread, and we use main thread mostly for user code

File: src/ray/raylet/node_manager.cc
Comment (clarng) [reply src/ray/raylet/node_manager.cc:825]: confirmed we will do this in follow up

File: src/ray/raylet/node_manager.cc
Comment (clarng) [reply src/ray/raylet/node_manager.cc:2996]: confirmed we will do this in follow up

**Comment by clarng** [comment]
> Looks like we only handle OOM case in this PR. Are you also planning to follow up with other types of failures (e.g., worker failure)?

sync'ed offline - we will handle other failures in a follow up PR

File: src/ray/raylet/node_manager.cc
Comment (clarng) [reply src/ray/raylet/node_manager.cc]: per offline discussion we are keeping track of the failures in task manager, so the state api can query it

File: src/ray/core_worker/task_manager.cc
Comment (stephanie-wang) [inline src/ray/core_worker/task_manager.cc:437]: @rkooo567 I'm confused, why do we need to save this? We should already be storing the exception as the task's return value, so why do we need the additional cache for failure reasons?

IMO, it would be better to only have this in the raylet, which can flush worker errors to the GCS worker table. If the state API needs that info, it can join the worker table info with the task table info.

File: src/ray/common/ray_config_def.h
Comment (stephanie-wang) [inline src/ray/common/ray_config_def.h:93]: Can we only add this setting instead of both? Even better would be if we could reuse an existing setting, but I'm not sure if there's one that would fit.

File: python/ray/exceptions.py
Comment (stephanie-wang) [inline python/ray/exceptions.py:381]: Can we improve this error message? It should include information similar to the message that we print on the raylet.

File: python/ray/exceptions.py
Comment (stephanie-wang) [inline python/ray/exceptions.py:393]: Same here.

File: python/ray/exceptions.py
Comment (stephanie-wang) [reply python/ray/exceptions.py:381]: Also, these should inherit from RayTaskError.

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (stephanie-wang) [reply src/ray/core_worker/transport/direct_task_transport.cc:593]: It would be best if we don't need to make the assumption that the code is using the same thread. I'm still a bit confused why that assumption exists.

Is there a way we can remove both that assumption and the check?

File: src/ray/protobuf/node_manager.proto
Comment (stephanie-wang) [inline src/ray/protobuf/node_manager.proto:343]: Can we use `ErrorType` instead of `RayException` in CoreWorker and Raylet?

Seems `RayException` is meant for exceptions in the language frontend, so we should try to keep the backend agnostic to these if possible.

**Review by stephanie-wang — CHANGES_REQUESTED** [review]

**Review by scv119 — COMMENTED** [review]
it's a bit unfortunate we need a separate path for this error.

File: python/ray/_private/test_utils.py
Comment (stephanie-wang) [inline python/ray/_private/test_utils.py:1429]:
```suggestion
# Send an RPC to the raylet to have it self-destruct its process.
```

File: python/ray/exceptions.py
Comment (stephanie-wang) [inline python/ray/exceptions.py:390]: Hmm wouldn't it be better to set the message string here and pass any related metadata, similar to what we do for other errors?

We could generate this error from multiple sources in the backend, so it will be harder to keep the message consistent if we do that. It's also a bit easier to read if we keep the message here.

File: src/ray/common/ray_config_def.h
Comment (stephanie-wang) [reply src/ray/common/ray_config_def.h:95]: Bump on this comment.

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (stephanie-wang) [inline src/ray/core_worker/transport/direct_task_transport.cc:587]: Consider moving this into a separate method for readability.

File: src/ray/raylet/node_manager.cc
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.cc:835]: I don't think this is quite at WARNING level, maybe DEBUG or INFO.

**Comment by clarng** [comment]
> it's a bit unfortunate we need a separate path for this error.

We proposed two options, either cancel the executing task and return the code via the push task response, or fetch through raylet via new rpc.

Seems the preference was to add new rpc, let me know if we should explore the cleaner option as well

File: src/ray/raylet/node_manager.cc
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.cc:2996]: Not sure that the raylet-level error message works as as the user-level exception message. I think we should probably tailor the OutOfMemoryError to the task/worker that died.

File: src/ray/raylet/node_manager.cc
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.cc:3023]: I think DEBUG is probably better here.

File: src/ray/raylet/node_manager.cc
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.cc:3026]: Seems safer not to assert here, since it should be a harmless race.

File: src/ray/protobuf/node_manager.proto
Comment (stephanie-wang) [inline src/ray/protobuf/node_manager.proto:338]: I think this proto is a bit misleading since it sounds like we're getting the return value of the task. Can we rename it to `GetTaskFailureCause` or something similar? Or maybe even `GetWorkerFailureCause` since that's what we're actually interested in (it could've been an actor that failed too).

File: python/ray/tests/test_memory_pressure.py
Comment (stephanie-wang) [inline python/ray/tests/test_memory_pressure.py:175]: This test probably belongs somewhere else since it doesn't have to do with memory monitoring. Maybe in a test_failure suite.

File: python/ray/tests/test_memory_pressure.py
Comment (stephanie-wang) [inline python/ray/tests/test_memory_pressure.py:183]: Can we also add a test that checks that we do receive this exception?

**Comment by stephanie-wang** [comment]
> it's a bit unfortunate we need a separate path for this error.

I think it is kind of unavoidable. Previously we had a longstanding TODO to try to get more info for system-level task errors, and it seems to require a new codepath to do it.

File: python/ray/exceptions.py
Comment (clarng) [reply python/ray/exceptions.py:390]: I looked at the existing way of formatting and have some concerns
- there is a lot of deserialization logic happening in the frontend and thus requiring a thick client, i.e. we will likely need to reimplement this logic in Java. The backend won't have to re-implement this code each time it tries to throw it, it can used a shared library, which is in c++ to format the message.
- In the future when we reformat the message we will have to change the payload on the wire / proto, i.e. its one of the reasons why we have this proto for propagating the error to the frontend. Wouldn't it be harder to maintain to have to keep adding new proto for each kind of error type?

```
message RayErrorInfo {
  oneof error {
    ActorDeathCause actor_died_error = 2;
    RuntimeEnvFailedContext runtime_env_setup_failed_error = 3;
    string error_message = 4;
  }
  // The type of error that caused the exception.
  optional ErrorType error_type = 11;
} 
```

Comment (stephanie-wang) [reply python/ray/exceptions.py:390]: I think it's fine to have a thicker client for error deserialization (it's not a performance issue and the error formatting is pretty low effort). Even if we have a helper method to re-implement in the backend, I'm a bit concerned it will be more brittle because we can forget to call it. In this case, having a more structured proto actually seems worth it since it is well-defined what metadata the backend is required to provide, and we can unit-test that behavior more easily than checking error message content.

That said, what metadata do we need to pass back from the NodeManager? Some of the fields like worker ID and node ID are probably already available on the driver side, so we can populate them there. For other metadata that's only available on the NodeManager side, we can pass this back as a shorter formatted error string that gets embedded in the user-level exception message.

Comment (clarng) [reply python/ray/exceptions.py:390]: How are we going to address the maintenance issue? Don't we need to re-implement and maintain this for each language like Python + Java?

To make this more robust we should add helper functions / factory method that generates the payload, i.e.

rpc::RayErrorInfo Errors::NewOutOfMemoryError(used_mem, total_mem, task_id, task_name, ......) {
     rpc::RayErrorInfo  error_info;
      std::stringstream worker_exit_message_ss;
          worker_exit_message_ss
              << "System memory low at node with IP " << latest_worker->IpAddress()
              << ". Used memory (" << used_bytes_gb << "GB) / total capacity ("
              << total_bytes_gb << "GB) (" << usage_fraction << ") exceeds threshold "
              << usage_threshold << ", killing latest task with name "
              << latest_worker->GetAssignedTask().GetTaskSpecification().GetName()
              << " and " << id_ss.str() << " to avoid running out of memory.\n"
              << "This may indicate a memory leak in a task or actor, or that too many "
                 "tasks are running in parallel.\n";
          std::string worker_exit_message = worker_exit_message_ss.str();
     error_info.error_message  = worker_exit_message_ss;
     return error_info
}

The message here is more flexible to change and more unit-testable. If we put it on the wire it is very hard to unit test since we don't have ways to test the raylet api directly, most such tests are done via the python test or mocked raylet client.

If we do it the other way where we populate parts of the message in the backend and then front it would looks something like this:


rpc::RayErrorInfo Errors::NewOutOfMemoryError(used_mem, total_mem, task_id, task_name, ......) {
     rpc::RayErrorInfo  error_info;
      std::stringstream worker_exit_message_ss;
          worker_exit_message_ss
              << "System memory low at node with IP " << latest_worker->IpAddress()
              << ". Used memory (" << used_bytes_gb << "GB) / total capacity ("
              << total_bytes_gb << "GB) (" << usage_fraction << ") exceeds threshold "
                 "tasks are running in parallel.\n";
          std::string worker_exit_message = worker_exit_message_ss.str();
     error_info.error_message  = worker_exit_message_ss;
     return error_info
}

and then in the frontend:

def new_oom_error(node_id, worker_id, backend_message) :
     error = OOMError()
     error.message =usage_threshold << 
     error.message = f"node id {node_id} ..... {backend_message killing latest task with name "
             {get_task_name()} and {task_id} to avoid running out of memory. This may indicate a "
            "memory leak in a task or actor, or that too many "
     return error
              

Wouldn't it be harder to understand and maintain and test if we do it separately like this?

Comment (ericl) [reply python/ray/exceptions.py:390]: I don't see a strong reason to go either way. From the user perspective, it makes little difference, and the maintenance concerns seem OK in each solution.

It seems like a stylistic decision, so I'd vote to be consistent with how this is approached with other similar types of error messages.

File: src/ray/common/ray_config_def.h
Comment (clarng) [reply src/ray/common/ray_config_def.h:95]: can you clarify which setting to keep and which one to remove?

Comment (stephanie-wang) [reply src/ray/common/ray_config_def.h:95]: Er, I don't feel strongly about it, but semantically seems like just specifying a TTL would make sense. (I mainly just want to try to minimize how many settings we add to this config)

Comment (clarng) [reply src/ray/common/ray_config_def.h:95]: so do we hard code the period ? I checked but haven't found an existing setting that makes sense to piggy back on (if we do we might as well run the logic together with an existing periodic fn)

Comment (stephanie-wang) [reply src/ray/common/ray_config_def.h:95]: How about hard-coding the period to TTL * 2?

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (clarng) [reply src/ray/core_worker/transport/direct_task_transport.cc:596]: Thanks for catching!

File: python/ray/exceptions.py
Comment (clarng) [reply python/ray/exceptions.py:390]: @stephanie-wang  can you let me know what should be part of the error message generated by the raylet and which part of the error message should be constructed on the frontend? This helps me figure out what fields should be added to the proto for the message that needs to be introduced.

The current message we sent back for OOM is 

```

std::stringstream worker_exit_message_ss;
          worker_exit_message_ss
              << "System memory low at node with IP " << latest_worker->IpAddress()
              << ". Used memory (" << used_bytes_gb << "GB) / total capacity ("
              << total_bytes_gb << "GB) (" << usage_fraction << ") exceeds threshold "
              << usage_threshold << ", killing latest task with name "
              << latest_worker->GetAssignedTask().GetTaskSpecification().GetName()
              << " and " << id_ss.str() << " to avoid running out of memory.\n"
              << "This may indicate a memory leak in a task or actor, or that too many "
                 "tasks are running in parallel.\n";
          std::string worker_exit_message = worker_exit_message_ss.str();

          std::stringstream error_log_ss;
          error_log_ss
              << worker_exit_message
              << "To find the highest memory consumers, use `ray logs raylet.out -ip "
              << latest_worker->IpAddress() << "`.\n"
              << "Consider provisioning more memory on this node or reducing task "
                 "parallelism by requesting more CPUs per task. "
              << "To adjust the eviction threshold, set the environment variable "
                 "`RAY_memory_usage_threshold_fraction` when starting Ray. "
              << "To disable worker eviction, set the environment variable "
                 "`RAY_memory_monitor_interval_ms` to zero.";


```




The current message we sent back for NodeDied is 

```

RAY_LOG(DEBUG) << "Failed to fetch task result with status "
                     << get_task_failure_cause_reply_status.ToString()
                     << " node id: " << addr.raylet_id << " ip: " << addr.ip_address;

```

Comment (stephanie-wang) [reply python/ray/exceptions.py:390]: I think the only fields that aren't available on the driver already are the ones that have to do with physical resources on the node, like the current used memory or how long the task ran before it was killed. But I'll leave it up to you to decide which fields to send back, and we can always improve on this later.

The high-level bit as @ericl said is just to make sure that the error message is useful to the user. Right now, I feel like the OOM message doesn't really make sense as a task return error since the content was originally about logging the OOM kill event on the raylet, not explaining what happened for a specific task. The NodeDied error could also be more informative (explaining why it might've happened, what to do next, etc).

You could look at the various [ObjectLostErrors](https://github.com/ray-project/ray/blob/master/python/ray/exceptions.py#L371) for some examples of what sort of metadata and content to include.

Comment (clarng) [reply python/ray/exceptions.py:390]: I see we have various ways of generating the message and not a clear consistent way of doing things. For example, ActorDiedErrorContext has all the fields populated from the backend, ObjectStoreFullError / OutOfDiskError doesn't have fields propagated, just things hard-coded. ObjectLostError doesn't have fields propagated from the backend either - the object ref is already available in the frontend. So I am still a little confused, and would prefer not to leave things up to me. In the interest of time it is best if you tell me what you want to avoid more back and forth. And it would be appreciated if we don't spend cycles getting everything perfect now so we could move on to the other (equally) important things like making OOM handling work better. 

In the latest comment what I am hearing is that we should have a different error message, which was not the comment at the beginning of this thread. So it is best if you help me understand what you want done from the start to avoid the back and forth. Thanks!

Comment (clarng) [reply python/ray/exceptions.py:390]: @ericl   this PR is following 

> I don't see a strong reason to go either way. From the user perspective, it makes little difference, and the maintenance concerns seem OK in each solution.
> 
> It seems like a stylistic decision, so I'd vote to be consistent with how this is approached with other similar types of error messages.

Here is what we are doing today, we pass the message from the raylet through the task submitter

       ```
         } else if (reply.failure_type() ==
                           rpc::RequestWorkerLeaseReply::
                               SCHEDULING_CANCELLED_UNSCHEDULABLE) {
                  error_type = rpc::ErrorType::TASK_UNSCHEDULABLE_ERROR;
                  *(error_info.mutable_error_message()) =
                      reply.scheduling_failure_message();

```

Can you help me understand which part of this PR is not consistent ?

Comment (stephanie-wang) [reply python/ray/exceptions.py:390]: Looking through the current exceptions, it seems like most of them use the pattern of passing some error context to Python and doing most of the formatting there, so I would prefer if we can make this the norm. But I didn't intend for this to block the PR, so if it's easier to format in C++, we can stick with that for now and follow up later to make all the errors consistent.

However, can you address the comment about improving the user-facing error messages for OutOfMemoryError and NodeDiedError before we merge? Currently they sound more like system log messages instead of task/worker-specific errors.

File: python/ray/_private/serialization.py
Comment (stephanie-wang) [inline python/ray/_private/serialization.py:236]: By the way, are these methods used? Also wasn't sure why they return `RayActorErrors`?

File: python/ray/_private/serialization.py
Comment (stephanie-wang) [inline python/ray/_private/serialization.py:245]: Is this used?

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (stephanie-wang) [inline src/ray/core_worker/transport/direct_task_transport.cc:621]: Call this method directly as the callback instead of having it return a callback?

**Review by stephanie-wang — APPROVED** [review]

**Comment by stephanie-wang** [comment]
By the way, could you also add an example of the error the user would now see? Just example stdout from a driver after receiving an OutOfMemoryError. Thanks!

**Review by scv119 — APPROVED** [review]

File: python/ray/_private/serialization.py
Comment (scv119) [inline python/ray/_private/serialization.py:298]: unused code

Comment (clarng) [reply python/ray/_private/serialization.py:298]: Cleaned up in follow up PR

**Pull request merged**
